Refactor/better student funcs

Aierbote opened the pull request:
**COMING SOON**

not yet ready!

Aierbote committed:
refactor(business-logic): better getStudent()

replace ~~for-i~~ with the public function getStudentList()

spotting the correct student with `find()` method of array

Aierbote committed:
refactor(business-logic): better createStudent()

avoiding re-assigning on the parameter variables `name` and `lastName`

using `normalizeName()` inside the logic of the object litteral

avoid using `push()` method, using spread operator to update array
students with the new object litteral

using `email,` instead of `email: email`
using `lectures,` instead of `lectures: [...lectures]` (or using push()

Aierbote committed:
refactor(business-logic): better updateStudent()

avoiding re-assigning on the parameter variables `name` and `lastName`

using `normalizeName()` inside the logic of assigning new value or if null keep the previous value

better name for variable of the `oldStudent` which is going to be updated

Aierbote committed:
refactor(business-logic): better deleteStudent()

iterate and spotting the element with the desired id, all thanks to `filter()`
method

from the definition of callback argument inside filter, renaming the id between
curly braches, in order to tell apart `id` parameter of `deleteStudent()` and
the `id` of the student at each iteration of `filter()`

Aierbote commented:
⚠️  I think I need to double check them before even getting started with this approve

I have not tested the functions and I think it still requires to get some more `console.log()` inside some of them

Rizaaal commented on script.js:
```diff
@@ -198,14 +182,8 @@ const getStudentList = () => {
-  }
-  console.log(`id ${id} not found in students`);
-  return null;
+  const students = getStudentList();
```
In realtà dovremmo semplicemente disfarci della funzione `getStudentList()`, mi ricordo il discorso di usare correttamente la funzione che ritorna la lista se dobbiamo prendere un elemento specifico ma visto che secondo me andrebbe semplicemente rimossa allora questo lo farei come prima e poi eventualmente le funzioni `getstudentList` e `getRegisterList` le eliminiamo e basta.

Rizaaal replied:
Ok visto che adesso `getStudentList()` prende i dati dal localStorage per me va bene

Rizaaal reviewed (changes requested):
i left a review...

Aierbote committed:
fix(business-logic): bug with createStudent and localStorage

it's not fixed yet. I finally save a new user in localStorage but it sadly
overwrite always the same index in the item 'students' inside Storage...

Aierbote committed:
fix(business-logic): `createStudent` + `lastStudId` with localStorage

Aierbote committed:
fix(business-logic): `deleteStudent` + `getStudent` with localStorage

to be fair, I had to change the equality check for id in both `getStudent()`
and \'prevStudents.filter()`'s callback, to be sure parameter id (which
can be (number | string) is casted as string.

Move the declaration of `deleteStudent()` after `createStudent()`

Aierbote committed:
fix(business-logic): `updateStudent` with localStorage

Move the declaration of `updateStudent()` after `deleteStudent()`

Aierbote commented:
> [!IMPORTANT] 
> I think I finished it, let's review it again to see follow the same pattern in other CRUD

Rizaaal commented on script.js:
```diff
@@ -117,26 +114,85 @@ const deleteAttendance = ({ registerId, attendanceId }) => {
+  const foundStudent = getStudent(id);
 
+  if (!!foundStudent) {
+    const prevStudents = getStudentList() || [];
```
nella riga 153, `getStudent` dipende da `getStudentList`...
Quindi come mai nella riga 156 abbiamo bisogno di creare un array vuoto se il controllo si trova dentro un if dove controlliamo se lo studente esiste? Se la lista di studenti non fosse stata dichiarata, non avremmo potuto trovare lo studente giusto?

Per me il controllo dovrebbe andare dentro `getStudentList` direttamente

Aierbote replied:
Thanks for spotting this tiny paradox, I removed the short-circuit in line 156

Rizaaal commented on script.js:
```diff
@@ -117,26 +114,85 @@ const deleteAttendance = ({ registerId, attendanceId }) => {
+          email: newEmail || stud.email,
+          lectures: newLectures || stud.lectures
+        }
+        : { ...stud };
```
Ma creiamo un nuovo oggetto anzichè restituirlo per l'immutabilità giusto?

Aierbote replied:
Yes, I liked this style to show it gets the same properties

Rizaaal reviewed (commented):
Fantastico, ho solo qualche domanda

Aierbote committed:
refactor(js): remove little paradox in `deleteStudent`

Rizaaal reviewed (approved):
ok
